source/sdk/Buffer.ooc

```diff
@@ -0,0 +1,70 @@
+/*
+* Copyright (C) 2015 - Simon Mika <simon@mika.se>
+*
+* This sofware is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+*
+* This software is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this software. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+Buffer: cover {
+	_pointer: Pointer
+	_size: Int
+	pointer ::= this _pointer
+	size ::= this _size
+	init: func@ { this init(null, 0) }
+	init: func@ ~allocate (size: Int) { this init(malloc(size), size) }
+	init: func@ ~fromData (=_pointer, =_size)
+	free: func@ -> Bool {
+		result := this _pointer != null && this _size != 0
+		if (result) {
+			free(this _pointer)
+			this _pointer = null
+			this _size = 0
+		}
+		result
+	}
+	reset: func (value: Int = 0) { memset(this _pointer, value, this _size) }
+	slice: func ~untilEnd (start: Int) -> This {
+		this slice(start, this _size - start)
+	}
+	slice: func (start, distance: Int) -> This { // call by value -> modifies copy of cover
+		size := distance absolute
+		if (start < 0)
+			start = this _size + start
+		if (distance < 0)
+			start -= size
+		start < this size ? This new(this _pointer + start, Int minimum(size, this _size - start)) : This empty
```

Comment:
I think this function is incorrect. `This new(pointer + start, size)` will create a `Buffer` which will try to call `free()` on a pointer which was not allocated by `malloc`. This may crash or even worse, not crash immediately.
Another problem is when someone calls `buffer2 := buffer slice(0, ...)` and attempts to free both buffers (double free ofc).